connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala

```diff
@@ -132,12 +132,32 @@ class SparkConnectPlanner(session: SparkSession) {
    * wrap such fields into proto messages.
    */
   private def transformSample(rel: proto.Sample): LogicalPlan = {
+    val input = Dataset.ofRows(session, transformRelation(rel.getInput))
+    val plan = if (rel.getForceStableSort) {
+      // It is possible that the underlying dataframe doesn't guarantee the ordering of rows in its
+      // constituent partitions each time a split is materialized which could result in
+      // overlapping splits. To prevent this, we explicitly sort each input partition to make the
+      // ordering deterministic. Note that MapTypes cannot be sorted and are explicitly pruned out
+      // from the sort order.
+      val sortOrder = input.logicalPlan.output
+        .filter(attr => RowOrdering.isOrderable(attr.dataType))
+        .map(SortOrder(_, Ascending))
+      if (sortOrder.nonEmpty) {
+        Sort(sortOrder, global = false, input.logicalPlan)
+      } else {
+        input.logicalPlan
+      }
+    } else {
+      input.cache()
```

Comment:
why do we cache here?

Reply:
This follows the existing implementation: https://github.com/apache/spark/blob/e23983a32df2bdb2c25673bb4f2a99b3c319ad23/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L2572

Reply:
This code path is for `df.sample` as well, does it need cache too?

Reply:
I still feel like we should introduce a new proto which only try to make the input dataset deterministic (by sorting or caching),  to avoid mixing RandomSplit with Sample.

Reply:
> This code path is for df.sample as well, does it need cache too?

I think the original Sample is wrongly cached here

Reply:
@zhengruifeng can you add such a proto message?

Reply:
ok, let me have a try

Reply:
https://github.com/apache/spark/pull/39151

Reply:
Yes. We should avoid cached plan for sample.
@zhengruifeng You will fix this issue? or I try to fix it.

Reply:
@zhengruifeng what is the purpose of the new proto message? The deterministic part is handled inside the planner. Exposing it as a new message indicates there is used beyond the plan parameter?

Reply:
Then we need a new proto message for randomSplit, and can't reuse Sample, to avoid adding extra cache for `df.sample`.

Reply:
ok, please just fix cache for Sample @beliefer

Reply:
I submit another PR. refer https://github.com/apache/spark/pull/39182